src/cf_create.cpp

```diff
@@ -0,0 +1,331 @@
+#include "structs.hpp"
+#include "awake.hpp"
+#include "utils.hpp"
+#include "db.hpp"
+#include "comm.hpp"
+#include "interpreter.hpp"
+#include "handler.hpp"
+#include "utils.hpp"
+#include "screen.hpp"
+#include "constants.hpp"
+#include "olc.hpp"
+#include "newmagic.hpp"
+
+#define CH d->character
+#define FORM d->edit_obj
+
+#define CFEDIT_MENU 0
+#define CFEDIT_TYPE 1
+#define CFEDIT_NAME 2
+#define CFEDIT_RATING 3
+#define CFEDIT_WOUND 4
+
+extern int get_program_skill(char_data *ch, obj_data *prog, int target);
+
+#define COMPLEX_FORM_TYPES 16
+
+int complex_form_programs[COMPLEX_FORM_TYPES] = {
+  SOFT_ATTACK,
+  SOFT_SLOW,
+  SOFT_COMPRESSOR,
+  SOFT_DECRYPT,
+  SOFT_RELOCATE,
+  SOFT_SLEAZE,
+  SOFT_TRACK,
+  SOFT_ARMOR,
+  SOFT_CAMO,
+  SOFT_CRASH,
+  SOFT_DEFUSE,
+  SOFT_EVALUATE,
+  SOFT_VALIDATE,
+  SOFT_SWERVE,
+  SOFT_CLOAK,
+  SOFT_LOCKON
+};
+
+void cfedit_disp_menu(struct descriptor_data *d)
+{
+  CLS(CH);
+  send_to_char(CH, "1) Name: ^c%s^n\r\n", FORM->restring);
```

Comment:
Is the form guaranteed to always exist and have a restring? If so this is fine, but otherwise this will crash. `GET_OBJ_NAME(FORM)` handles these edge cases and is generally safer.

Reply:
By the time it reaches this time in the execution area, yes. But I'll change it anyways as it shouldn't matter.